### ISSUE TITLE
JOSS Review #2

### ISSUE THREAD

**Issue by pescap** [DIDSR/calzone#5]
This issue is related to [openjournals/joss-reviews/issues/8026](https://github.com/openjournals/joss-reviews/issues/8026).

Hi, very good work on such an important subject!

# Paper

- Pag.1, l.18: What is Se/Sp? Add ref for AUC, Se/Sp.
- In the paper you state “provides functions to compute various calibration metrics.” Here I would add more guidance and general context before listing them one by one — to help the reader understand that some metrics do X, others do Y, and so on.

Try to show in which ways you are extending and adding specialized features to the generic `scikit-learn` functionality. When should one use `calzone`? This could go in the statement of interest. Because again, I feel like one does not need to be an expert in calibration to need to go beyond `scikit-learn` and use `calzone`.

For an introduction to this subject, `scikit-learn` entry on [calibration](https://scikit-learn.org/stable/modules/calibration.html) good as an introduction. Maybe actually, the reader could start with that, and then get to why calzone, and the specificities you add. 

This boundary between: go with `scikit-learn` v/s use `calzone` is key to me. You mention it in the statement of need but I would be more specific. 

Also, mention other libraries and how you present a novelty and differentiation, e.g.: 
- [ml-calibration](https://github.com/apple/ml-calibration)
- [pycaleva](https://github.com/MartinWeigl/pycaleva) 

# Doc

- "calzone aims to assess whether a model achieves moderate calibration, meaning whether [...]". The notation for this equation is different from the one in *Statement of need*. Maybe use the same consistent notation across the docs and JOSS paper.

In [metrics_summary](https://calzone-docs.readthedocs.io/en/latest/notebooks/metrics_summary.html):  
- The table is interesting, but with the ReadTheDocs default zoom I cannot read it well, it appears too small. Maybe turn the table into a Markdown table.

In general, for the docs, I would be more "guiding".

- Maybe add references or small explanations for a non-expert audience. For example:  

```
To accurately assess the calibration of machine learning models, it is essential to have a comprehensive and 
representative dataset with sufficient coverage of the prediction space. 
The calibration metric is not meaningful if the dataset is not representative of the true intended population.
```

and  

```
If you observe general over- or under-estimation of probabilities for a given class, 
consider applying a prevalence adjustment to determine if it’s solely due to a prevalence 
shift. After prevalence adjustment, plot the reliability diagrams again and examine 
the results of calibration metrics.
```

"Representative dataset with sufficient coverage": explain a bit more.  
I do not master the definitions like "prevalence", "prevalence adjustment" (which type, what).
Or maybe refer to the dedicated section in the docs.

Check typos: *representative*, *alternatively*.

[reliability_diagram](https://calzone-docs.readthedocs.io/en/latest/notebooks/reliability_diagram.html): check typos. Give more details. For example, detail what is written as comments:  

```
### We will artificially drop the prevalence of class 1
# The top-class reliability diagram will still look good
# But the class-1 reliability diagram will be very bad
```

Give a detail of the different kinds of methods here. Because afterwards, in the next sections, you detail the different tests that have diverse usages. I would summarize the broad usage of all at the beginning and also in the paper. My impression when I started with the dict was that there were a lot of things measured, but not a clear view of the whole. Maybe be more consistent in explaining the example data too.

[Validating metrics](https://calzone-docs.readthedocs.io/en/latest/notebooks/validation.html): add these tests (or some of them) to the unit tests?

# Curiosity

What about unbalanced datasets? Maybe elaborate on this.

**Comment by pescap** [comment]
> This issue is related to [openjournals/joss-reviews/issues/8026](https://github.com/openjournals/joss-reviews/issues/8026).
> 
> Hi, very good work on such an important subject!
> 
> # Paper
> * Pag.1, l.18: What is Se/Sp? Add ref for AUC, Se/Sp.
> * In the paper you state “provides functions to compute various calibration metrics.” Here I would add more guidance and general context before listing them one by one — to help the reader understand that some metrics do X, others do Y, and so on.
> 
> # Doc
> * "calzone aims to assess whether a model achieves moderate calibration, meaning whether [...]". The notation for this equation is different from the one in _Statement of need_. Maybe use the same consistent notation across the docs and JOSS paper.
> 
> In [metrics_summary](https://calzone-docs.readthedocs.io/en/latest/notebooks/metrics_summary.html):
> 
> * The table is interesting, but with the ReadTheDocs default zoom I cannot read it well, it appears too small. Maybe turn the table into a Markdown table.
> 
> In general, for the docs, I would be more "guiding".
> 
> * Maybe add references or small explanations for a non-expert audience. For example:
> 
> ```
> To accurately assess the calibration of machine learning models, it is essential to have a comprehensive and 
> representative dataset with sufficient coverage of the prediction space. 
> The calibration metric is not meaningful if the dataset is not representative of the true intended population.
> ```
> 
> and
> 
> ```
> If you observe general over- or under-estimation of probabilities for a given class, 
> consider applying a prevalence adjustment to determine if it’s solely due to a prevalence 
> shift. After prevalence adjustment, plot the reliability diagrams again and examine 
> the results of calibration metrics.
> ```
> 
> "Representative dataset with sufficient coverage": explain a bit more. I do not master the definitions like "prevalence", "prevalence adjustment" (which type, what). Or maybe refer to the dedicated section in the docs.
> 
> Check typos: _representative_, _alternatively_.
> 
> [reliability_diagram](https://calzone-docs.readthedocs.io/en/latest/notebooks/reliability_diagram.html): check typos. Give more details. For example, detail what is written as comments:
> 
> ```
> ### We will artificially drop the prevalence of class 1
> # The top-class reliability diagram will still look good
> # But the class-1 reliability diagram will be very bad
> ```
> 
> Give a detail of the different kinds of methods here. Because afterwards, in the next sections, you detail the different tests that have diverse usages. I would summarize the broad usage of all at the beginning and also in the paper. My impression when I started with the dict was that there were a lot of things measured, but not a clear view of the whole. Maybe be more consistent in explaining the example data too.
> 
> [Validating metrics](https://calzone-docs.readthedocs.io/en/latest/notebooks/validation.html): maybe add these tests (or some of them) to the unit tests?
> 
> Finally: Maybe try to show in which ways you are extending and adding specialized features to the generic scikit-learn functionality. When should one use `calzone`? This could go in the statement of interest. Because again, I believe one does not need to be an expert in calibration to need to go beyond scikit-learn and use `calzone`.
> 
> For a good introduction, `scikit-learn` entry on [calibration](https://scikit-learn.org/stable/modules/calibration.html) is very good as an introduction. Maybe actually, reader could start with that, and then get to why calzone, and the specificities you add.
> 
> This boundary between: go with `scikit-learn` v/s use `calzone` is key to me.
> 
> # Curiosity
> What about unbalanced datasets? I understand that there is a very strong connection between loss of calibration and unbalanced datasets. Maybe elaborate on this.

**Comment by fabian-s** [comment]
@qiancao while we wait for the other reviewer, please start adressing these points

**Comment by qiancao** [comment]
Thank you for your detailed review of Calzone. We've been making updates to the manuscript and the code to address your comments. The revised manuscript is [here](https://github.com/DIDSR/calzone/blob/main/paper/paper.pdf).

> This issue is related to [openjournals/joss-reviews/issues/8026](https://github.com/openjournals/joss-reviews/issues/8026).
> 
> Hi, very good work on such an important subject!

> # Paper
> * Pag.1, l.18: What is Se/Sp? Add ref for AUC, Se/Sp.

We have revised this paragraph to define AUC, Se, and Sp as area under ROC curve, sensitivity and specificity, respectively.

> * In the paper you state “provides functions to compute various calibration metrics.” Here I would add more guidance and general context before listing them one by one — to help the reader understand that some metrics do X, others do Y, and so on.
> 
> Try to show in which ways you are extending and adding specialized features to the generic `scikit-learn` functionality. When should one use `calzone`? This could go in the statement of interest. Because again, I feel like one does not need to be an expert in calibration to need to go beyond `scikit-learn` and use `calzone`.
> 
> For an introduction to this subject, `scikit-learn` entry on [calibration](https://scikit-learn.org/stable/modules/calibration.html) good as an introduction. Maybe actually, the reader could start with that, and then get to why calzone, and the specificities you add.
> 
> This boundary between: go with `scikit-learn` v/s use `calzone` is key to me. You mention it in the statement of need but I would be more specific.
> 
> Also, mention other libraries and how you present a novelty and differentiation, e.g.:
> 
> * [ml-calibration](https://github.com/apple/ml-calibration)
> * [pycaleva](https://github.com/MartinWeigl/pycaleva)

Calzone aims to be a comprehensive library for the evaluation of probability calibration. We have revised the introduction section to differentiate Calzone from existing packages, see lines 30-45:

While existing libraries such as scikit-learn include basic tools like reliability diagrams and expected calibration error, they lack support for more comprehensive and flexible evaluation metrics—such as reliability diagrams with error bars, class-conditional calibration error, different binning schemes, or statistical significance testing for miscalibration. This is also the case with other calibration-focused libraries, such as ml-calibration, uncertainty-toolbox, and pycaleva. For example, ml-calibration provides advanced controls for plotting reliability diagrams and computing smooth expected calibration error but does not include statistical tests for miscalibration [@ml-calibration]. The uncertainty-toolbox focuses on calibration methods rather than assessment [@uncertaintyToolbox]. The pycaleva package overlaps with many functionalities in calzone, but it does not support Cox's calibration analysis, Wald intervals for reliability, or custom curve fitting methods for expected calibration error [@pycaleva]. In contrast, Calzone emphasizes the evaluation of calibration. It features a comprehensive set of calibration metrics, statistical tests (e.g., hypothesis testing for miscalibration), and visualization tools tailored for many types of classification tasks (e.g., multi-class metrics). The package is designed to help users not only visualize miscalibration but also quantify and statistically validate it in a consistent and interpretable way.

> # Doc
> * "calzone aims to assess whether a model achieves moderate calibration, meaning whether [...]". The notation for this equation is different from the one in _Statement of need_. Maybe use the same consistent notation across the docs and JOSS paper.

We've change the notation in the documentation to make it consistent with that of the paper.

> In [metrics_summary](https://calzone-docs.readthedocs.io/en/latest/notebooks/metrics_summary.html):
> 
> * The table is interesting, but with the ReadTheDocs default zoom I cannot read it well, it appears too small. Maybe turn the table into a Markdown table.
>

Thank you for your comments. We've changed the table to markdown.

> In general, for the docs, I would be more "guiding".
> 
> * Maybe add references or small explanations for a non-expert audience. For example:
> 
> ```
> To accurately assess the calibration of machine learning models, it is essential to have a comprehensive and 
> representative dataset with sufficient coverage of the prediction space. 
> The calibration metric is not meaningful if the dataset is not representative of the true intended population.
> ```
> 
> and
> 
> ```
> If you observe general over- or under-estimation of probabilities for a given class, 
> consider applying a prevalence adjustment to determine if it’s solely due to a prevalence 
> shift. After prevalence adjustment, plot the reliability diagrams again and examine 
> the results of calibration metrics.
> ```
> 
> "Representative dataset with sufficient coverage": explain a bit more. I do not master the definitions like "prevalence", "prevalence adjustment" (which type, what). Or maybe refer to the dedicated section in the docs.

Thank you for your comments. We've described prevalence adjustment in our documentation: https://calzone-docs.readthedocs.io/en/latest/notebooks/prevalence_adjustment.html. Additionally, we expanded our discussion of prevalence adjustment in the paper. See lines 123-128:

    Calzone offers prevalence adjustment to correct for differences in disease prevalence between training and testing data. Calibration is based on posterior probability, so a shift in prevalence can cause miscalibration. The adjusted probability is calculated as: ... directly if known [@weijie_prevalence_adjustment;@prevalence_shift;@gu_likelihod_ratio;@Prevalence_HORSCH].

> Check typos: _representative_, _alternatively_.
> 
> [reliability_diagram](https://calzone-docs.readthedocs.io/en/latest/notebooks/reliability_diagram.html): check typos. Give more details. For example, detail what is written as comments:
> 
> ```
> ### We will artificially drop the prevalence of class 1
> # The top-class reliability diagram will still look good
> # But the class-1 reliability diagram will be very bad
> ```
> 
> Give a detail of the different kinds of methods here. Because afterwards, in the next sections, you detail the different tests that have diverse usages. I would summarize the broad usage of all at the beginning and also in the paper. My impression when I started with the dict was that there were a lot of things measured, but not a clear view of the whole. Maybe be more consistent in explaining the example data too.
>

Thank you for your comments. We added a summary of metrics in the beginning and in the quick start sections of documentation page.

> 
> [Validating metrics](https://calzone-docs.readthedocs.io/en/latest/notebooks/validation.html): add these tests (or some of them) to the unit tests?
>

Thank you for your comments. We have added two more tests to make sure the result is reasonable.

> # Curiosity
> What about unbalanced datasets? Maybe elaborate on this.

Here, we interpret an "unbalanced dataset" to be a dataset where the proportions of target classes are not similar. In a binary classification problem, this might mean that negative class far outnumber the positive class (i.e. low prevalence). In Calzone, we provide users with the ability to do prevalence adjustment. We have provided more information in the documentation: https://calzone-docs.readthedocs.io/en/latest/notebooks/prevalence_adjustment.html
